### PR TITLE
Impl vdso for time accuracy and less ocalls

### PR DESCRIPTION
**Pull request by ClawSeven**

### CODE REVIEW

File: src/libos/src/lib.rs [hunk 1]
```diff
@@ -78,8 +82,6 @@ use crate::process::pid_t;
 
 #[macro_use]
 mod prelude;
-#[macro_use]
-mod error;
```
Comment (volcano0dr) [inline src/libos/src/lib.rs:82]: You can export the errno crate to error mod.

Comment (ClawSeven) [reply src/libos/src/lib.rs]: Done

File: src/libos/src/error/mod.rs [hunk 2]
```diff
@@ -1,69 +1,3 @@
 use super::*;
```
Comment (volcano0dr) [inline src/libos/src/error/mod.rs:1]: remove `use super::*;`

File: src/libos/src/prelude.rs [hunk 3]
```diff
@@ -19,6 +19,7 @@ pub use crate::error::*;
 pub use crate::fs::{File, FileDesc, FileRef};
 pub use crate::process::{pid_t, uid_t};
 pub use crate::util::sync::RwLock;
+pub use vdso_time::ClockId;
```
Comment (volcano0dr) [inline src/libos/src/prelude.rs:22]: remove `pub use vdso_time::ClockId`.

File: src/libos/src/time/up_time.rs [hunk 4]
```diff
@@ -1,11 +1,12 @@
-use super::{do_clock_gettime, ClockID};
+use super::do_clock_gettime;
```
Comment (volcano0dr) [inline src/libos/src/time/up_time.rs:1]: Don't need to change.
`use super::{do_clock_gettime, ClockId};`

File: src/libos/src/process/syscalls.rs [hunk 5]
```diff
@@ -11,9 +11,10 @@ use super::process::ProcessFilter;
 use super::spawn_attribute::{clone_spawn_atrributes_safely, posix_spawnattr_t, SpawnAttr};
 use crate::prelude::*;
 use crate::syscall::CpuContext;
-use crate::time::{timespec_t, ClockID};
+use crate::time::timespec_t;
```
Comment (volcano0dr) [inline src/libos/src/process/syscalls.rs:14]: `use crate::time::{timespec_t, ClockId};`

File: src/libos/src/process/do_futex.rs [hunk 6]
```diff
@@ -4,7 +4,8 @@ use std::intrinsics::atomic_load_seqcst;
 use std::sync::atomic::{AtomicBool, Ordering};
 
 use crate::prelude::*;
-use crate::time::{timespec_t, ClockID};
+use crate::time::timespec_t;
```
Comment (volcano0dr) [inline src/libos/src/process/do_futex.rs:7]: `use crate::time::{timespec_t, ClockId};`

File: src/libos/src/process/do_exec.rs [hunk 7]
```diff
@@ -102,12 +102,13 @@ pub fn do_exec(
 // Blocking wait until there is only one thread in the calling process
 fn wait_for_other_threads_to_exit(current_ref: &ThreadRef) {
     use super::do_futex::{self, FutexTimeout};
-    use crate::time::{timespec_t, ClockID};
+    use crate::time::timespec_t;
```
Comment (volcano0dr) [inline src/libos/src/process/do_exec.rs:105]: `use crate::time::{timespec_t, ClockId};`

File: src/libos/src/fs/timer_file.rs [hunk 8]
```diff
@@ -1,6 +1,6 @@
 use super::*;
 
-use crate::time::{clockid_t, itimerspec_t, timespec_t, ClockID};
+use crate::time::{clockid_t, itimerspec_t, timespec_t};
```
Comment (volcano0dr) [inline src/libos/src/fs/timer_file.rs:3]: `use crate::time::{clockid_t, itimerspec_t, timespec_t, ClockId};`

File: src/libos/src/fs/syscalls.rs [hunk 9]
```diff
@@ -6,7 +6,7 @@ use super::file_ops::{
 };
 use super::fs_ops;
 use super::fs_ops::{MountFlags, MountOptions, UmountFlags};
-use super::time::{clockid_t, itimerspec_t, timespec_t, timeval_t, ClockID};
+use super::time::{clockid_t, itimerspec_t, timespec_t, timeval_t};
```
Comment (volcano0dr) [inline src/libos/src/fs/syscalls.rs:9]: `use super::time::{clockid_t, itimerspec_t, timespec_t, timeval_t, ClockId};`

File: src/libos/src/error/mod.rs (hunk 2 above)
Comment (ClawSeven) [reply src/libos/src/error/mod.rs]: Done

File: src/libos/src/prelude.rs (hunk 3 above)
Comment (ClawSeven) [reply src/libos/src/prelude.rs]: Done

File: src/libos/src/time/up_time.rs (hunk 4 above)
Comment (ClawSeven) [reply src/libos/src/time/up_time.rs]: Done

File: src/libos/src/process/syscalls.rs (hunk 5 above)
Comment (ClawSeven) [reply src/libos/src/process/syscalls.rs]: Done

File: src/libos/src/process/do_futex.rs (hunk 6 above)
Comment (ClawSeven) [reply src/libos/src/process/do_futex.rs]: Done

File: src/libos/src/process/do_exec.rs (hunk 7 above)
Comment (ClawSeven) [reply src/libos/src/process/do_exec.rs]: Done

File: src/libos/src/fs/timer_file.rs (hunk 8 above)
Comment (ClawSeven) [reply src/libos/src/fs/timer_file.rs]: Done

File: src/libos/src/fs/syscalls.rs (hunk 9 above)
Comment (ClawSeven) [reply src/libos/src/fs/syscalls.rs]: Done

File: src/libos/src/exception/cpuid.rs [hunk 10]
```diff
@@ -218,6 +220,17 @@ impl CpuId {
         };
         cpuid_result
     }
+
+    pub fn support_sgx2(&self) -> bool {
+        if rsgx_get_enclave_mode() != EnclaveMode::Hw {
```
Comment (volcano0dr) [inline src/libos/src/exception/cpuid.rs:225]: remove `rsgx_get_enclave_mode` call.

File: src/libos/src/time/mod.rs [hunk 11]
```diff
@@ -334,3 +309,10 @@ impl itimerspec_t {
         Ok(())
     }
 }
+
+fn is_enable_vdso() -> bool {
+    if rsgx_get_enclave_mode() == EnclaveMode::Sim {
```
Comment (volcano0dr) [inline src/libos/src/time/mod.rs:314]:
```
match rsgx_get_enclave_mode() {
    EnclaveMode::Hw if is_cpu_support_sgx2() => ture,
    EnclaveMode::Sim => true,
    _ => false,
}
````

Add `time::init() `function.

```
pub fn init() {
     // init static var `IS_ENABLE_VDSO`;
     // init up_time;
}
```

Comment (ClawSeven) [reply src/libos/src/time/mod.rs]: Done

File: src/libos/src/exception/cpuid.rs (hunk 10 above)
Comment (ClawSeven) [reply src/libos/src/exception/cpuid.rs]: Done

File: src/libos/src/time/mod.rs [hunk 12]
```diff
@@ -1,9 +1,12 @@
+use self::exception::is_cpu_support_sgx2;
```
Comment (volcano0dr) [inline src/libos/src/time/mod.rs:1]: `use crate::exception::is_cpu_support_sgx2;`

File: src/libos/src/time/mod.rs [hunk 13]
```diff
@@ -27,6 +31,28 @@ pub type clock_t = i64;
...
+
+fn init_vdso() {
+    IS_ENABLE_VDSO.call_once(|| match rsgx_get_enclave_mode() {
+        EnclaveMode::Hw if is_cpu_support_sgx2() => true,
+        EnclaveMode::Sim => true,
+        _ => false,
+    });
+}
+
+fn is_enable_vdso() -> bool {
```
Comment (volcano0dr) [inline src/libos/src/time/mod.rs:49]: add `#[inline]`

File: src/libos/src/time/mod.rs [hunk 14]
```diff
@@ -27,6 +31,28 @@ pub type clock_t = i64;
...
+    match IS_ENABLE_VDSO.get() {
+        Some(enable_vdso) => *enable_vdso,
```
Comment (volcano0dr) [inline src/libos/src/time/mod.rs:51]: `IS_ENABLE_VDSO.get().map_or(false, |is_enable| *is_enable)`

File: src/libos/crates/vdso-time/src/lib.rs [hunk 15]
```diff
@@ -0,0 +1,622 @@
...
+        }
+        return_errno!(EINVAL, "Unsupported clock_mode");
+    }
+}
+
+unsafe impl Sync for Vdso {}
+unsafe impl Send for Vdso {}
+
+lazy_static! {
+    static ref VDSO: Option<Vdso> = Vdso::new().map_or_else(
```
Comment (volcano0dr) [inline src/libos/crates/vdso-time/src/lib.rs:411]: `static ref VDSO: Option<Vdso> = Vdso::new().ok();`

**Pull request merged**